core/src/main/java/com/linecorp/armeria/common/stream/StreamWriterOutputStream.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpData;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+final class StreamWriterOutputStream<T> extends OutputStream {
+
+    private final StreamWriter<T> writer;
+    private final Function<? super HttpData, ? extends T> httpDataConverter;
+    private ByteBuf buffer;
+    private final int maxBufferSize;
+    private boolean closed;
+
+    StreamWriterOutputStream(StreamWriter<T> writer,
+                             Function<? super HttpData, ? extends T> httpDataConverter,
+                             int maxBufferSize) {
+        requireNonNull(writer, "writer");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        checkArgument(maxBufferSize > 0, "maxBufferSize: %s (expected: > 0)", maxBufferSize);
+        this.writer = writer;
+        this.httpDataConverter = httpDataConverter;
+        this.maxBufferSize = maxBufferSize;
+        buffer = Unpooled.buffer(maxBufferSize, maxBufferSize);
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        ensureOpen();
+        buffer.writeByte(b);
+        maybeDrain();
+    }
```

Comment:
>  I don't see your implementation never blocks when a user calls `StreamWriterOutputStream.write().` 
> Could you make sure to wait until the previous `flush()` is `consumed` by the consumer by using `StreamMessageWriter.whenConsumed()`?

oh.. I think `StreamWriterOutputStream.write()` is **blocking call**.

Also we expect that user calls `write()&flush()` in a same thread with blocking manner,
So I think we don't have to wait until the previous `flush()` is consumed during write() 🤔

Can you tell me more about this idea? thanks!

Reply:
Sorry about the late review 😅 

I can't speak for trustin, but this is my thought:

I guess the expected behavior for our `StreamWriterOutputStream` is that `write` operations block until the object has actually been written.

Otherwise, users can continuously call `OutputStream#write`, and the data might be buffered somewhere in memory if the downstream reactive streams implementation doesn't process data quickly enough.
```
while (true) {
    StreamWriterOutputStream.write(bytes); // if the data isn't consumed quickly enough, the data will be buffered in memory
}
```

We can implement a concept of backpressure by making `OutputStream#write` blocking.
e.g.
```java
CompletableFuture<Void> writeFuture = CompletableFuture.completedFuture(null);
    @Override
    public void write(int b) throws IOException {
        writeFuture.join();
        // try to write
        if (!writer.tryWrite(httpDataConverter.apply(HttpData.wrap(buffer)))) {
            throw new IOException("Stream closed");
        }
        writeFuture = writer.whenConsumed();
    }
```

Reply:
Alternatively, we may want to give users some freedom to not consider backpressure.
In this case, it may be a better idea to consider performing the `StreamWriter#whenConsumed` check whenever the buffer is drained.

```
CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
    private void drain() throws IOException {
        if (!buffer.isReadable() || !writer.isOpen()) {
            return;
        }
        future.join();
        if (!writer.tryWrite(httpDataConverter.apply(HttpData.wrap(buffer)))) {
            throw new IOException("Stream closed");
        }
        buffer = Unpooled.buffer(maxBufferSize, maxBufferSize);
        future.whenConsumed();
    }
```

Reply:
(FYI) I'll check your comment until tomorrow morning 🙇